Add unit tests for cams_call.py controller

mosesowino opened the pull request:
- Added tests for the `attendance` function covering:
  empty request body,
   invalid JSON input,
  processing RealTime attendance logs,
  processing PunchLog entries,
- Mocked dependencies to isolate the endpoint logic
- Ensured proper calls to `handle_attendance_log` and `handle_punch_logs`
- Verified correct HTTP response for all cases

Commit:
- Added tests for the `attendance` function covering:
  empty request body,
   invalid JSON input,
  processing RealTime attendance logs,
  processing PunchLog entries,
- Mocked dependencies to isolate the endpoint logic
- Ensured proper calls to `handle_attendance_log` and `handle_punch_logs`
- Verified correct HTTP response for all cases

maniamartial commented on navari_cams_biometric/cams_biometric/controllers/test_cams_call.py:
```diff
@@ -0,0 +1,126 @@
+)
+
+
+class TestCamsCall(TestCase):
```
Good use of unittest(assuming you are new to frappe world), but since you we are using frappe framework, as we go forward, stick to FrappeTestCase.

Check on this sample.
kenya_compliance/kenya_compliance/doctype/navari_kra_etims_settings/test_navari_kra_etims_settings.py

mosesowino replied:
Switched from unittest to FrappeTestCase

maniamartial commented on navari_cams_biometric/cams_biometric/controllers/test_cams_call.py:
```diff
@@ -0,0 +1,126 @@
+        response = attendance()
+        self.assertEqual(response.status_code, 200)
+        mock_handle_punch_logs.assert_called_once_with("STG001", data_punchlog["PunchLog"]["Log"])
+
```
Add test: verify update_last_sync_time():

mosesowino replied:
added the test, please check.

maniamartial commented on navari_cams_biometric/cams_biometric/controllers/test_cams_call.py:
```diff
@@ -0,0 +1,126 @@
+            "InputType": "Fingerprint"
+        }]
+
+        mock_parser.parse.return_value.strftime.return_value = "2024-05-01 08:00:00"
```
I like the way you are doing isolation.
However, too much isolation is harmful.

When you mock too much, your tests no longer test real behavior. Instead, you're verifying that mocks interact in a certain way, not that the actual system works.

Example: If you mock frappe.get_doc().insert() and frappe.db.commit(), you aren't testing whether a Checkin is actually created in the database, only that your code tried to do so.

maniamartial replied:
If you decided to do so, make sure your mocks as dynamic as possible to avoid False Positive whereby test passes even if the production code is broken.

eg Assuming we have this function

```
def insert_checkin(employee, timestamp, log_type):
    doc = frappe.new_doc("Employee Checkin")
    doc.employee = employee
    doc.time = timestamp
    doc.log_type = log_type
    doc.insert()
    return doc
```


Then the dynamic test should be something like

```
def test_insert_checkin_creates_doc():
    employee = create_test_employee()  # helper to make function shorter.
    timestamp = frappe.utils.now()
    log_type = "IN"

    checkin = insert_checkin(employee.name, timestamp, log_type)

    assert checkin.doctype == "Employee Checkin"
    assert checkin.employee == employee.name
    assert checkin.log_type == "IN"
    assert frappe.db.exists("Employee Checkin", checkin.name)

```

mosesowino replied:
I’ve managed to rewrite the tests with minimal mocking to allow actual read/write interactions with the database.

However, I’ve just noticed that the file is now missing from my fork. Should I go ahead and recreate the file and push the changes again, or would you prefer I wait for further instructions?

maniamartial replied:
You can recreate

mosesowino replied:
please review the changes.

maniamartial commented:
Otherwise good test for beginner.

Commit:
Addressed some review comments:- Use frappeTestCase instead of unittest   -Add test case for update_last_sync_time()

Commit:
test: add integration-style tests for attendance and shift sync logic

- Avoided over-mocking to ensure database operations (e.g. Employee Checkin creation) are tested end-to-end
- Verified real insertion of Checkin records for handle_attendance_log and handle_punch_logs
- Tested attendance() controller with patching only for outer dependencies (frappe request, form_dict)
